File: material/fizz-buzz/fizz-buzz.go
package main

import (
	"fmt"
	"strconv"
)

func newCounter() func() int {
	ct := 0
	return func() int {
		ct++
		return ct
	}
}

func fizzBuzz(n int) string {
	if n%3 == 0 && n%5 == 0 {
		return "Fizz Buzz"
	} else if n%3 == 0 {
		return "Fizz"
	} else if n%5 == 0 {
		return "Buzz"
	}
	return strconv.Itoa(n)
}

func main() {
	max := 30
	c := newCounter()
	for n := c(); ; n = c() {
		if n < max {
			fmt.Print(fizzBuzz(n), ", ")
		} else {
			fmt.Println(fizzBuzz(n))
			break
		}
	}
}

File: material/fizz-buzz/fizz-buzz.cpp
#include <iostream>
#include <string>

auto newCounter() {
	return [ct=0]() mutable {
		return ++ct;
    };
}

auto fizzBuzz(auto n) {
	if (n%3==0 && n%5==0) {
		return std::string("Fizz Buzz");
	} else if (n%3==0) {
		return std::string("Fizz");
	} else if (n%5==0) {
		return std::string("Buzz");
	}
	return std::to_string(n);
}

int main() {
	auto max = 30;
	auto c = newCounter();
	for (auto n = c(); ; n = c()) {
		if (n < max) {
			std::cout << fizzBuzz(n) << ", ";
		} else {
			std::cout << fizzBuzz(n) << std::endl;
			break;
		}
	}
	return 0;
}
